### PR TITLE
add support for Tuya pink version of miflora

### LINKED ISSUE #2399: add support for Tuya pink version of miflora
**Describe the problem you have/What new integration you would like**

A new version of the xiaomi miflora exist, the pink version, that is incompatible with the old version but adds updated hardware and tuya support.
https://smarthomescene.com/reviews/xiaomi-miflora-plant-sensor-tuya-version-hhccjcy10-review/
I've pushed here an PullRequest that implement the new protocol.
https://github.com/esphome/esphome/pull/5402

More details in the PR, but the integration is easy enough to follow. Working great since 2 weeks now

**Please describe your use case for this integration and alternatives you've tried:**

**Additional context**

### PR DESCRIPTION
**Pull request by fariouche**
# What does this implement/fix?
Hello.
Added support for new miflora Pink (Tuya version) of the sensor.
This version is not compatible with the xiaomi version (white and green) and is using a BLE5.0 chip.
This is a simple component, tested and working (I have 3).
Parsing is inspired by the openmqttgateway (theengs decoder)
The code name is hhccjcy10 (see the last number is not 01 as the old one)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** Closes https://github.com/esphome/feature-requests/issues/2399

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3465

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
 - platform: xiaomi_hhccjcy10
    mac_address: 'xx:xx:xx:xx:xx:xx'
    temperature:
      name: "xiaomi_hhccjcy10 Temperature"
    moisture:
      name: "xiaomi_hhccjcy10 Moisture"
    illuminance:
      name: "xiaomi_hhccjcy10 Illuminance"
    conductivity:
      name: "xiaomi_hhccjcy10 Soil Conductivity"
    battery_level:
      name: "xiaomi_hhccjcy10 Battery Level"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by fariouche** [comment]
what do you think?

**Review by jesserockz — CHANGES_REQUESTED** [review]
> what do you think?

Code looks fine on quick glance, but PRs with failing CI, or labels like `needs-docs` or `needs-tests` dont usually get reviewed until resolved =)

**Comment by fariouche** [comment]
oh I see, thanks. I will check the docs and tests then thanks

**Comment by fariouche** [comment]
All green now :smile: 
Thanks!

File: esphome/components/xiaomi_hhccjcy10/sensor.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    UNIT_LUX,
+    CONF_CONDUCTIVITY,
+    UNIT_MICROSIEMENS_PER_CENTIMETER,
+    ICON_FLOWER,
+    DEVICE_CLASS_BATTERY,
+    CONF_BATTERY_LEVEL,
+)
+
+DEPENDENCIES = ["esp32_ble_tracker"]
+AUTO_LOAD = ["xiaomi_ble"]
```
Comment (jesserockz) [inline esphome/components/xiaomi_hhccjcy10/sensor.py]: As this doesnt use the xiaomi_ble code, this is not required (paired with below removal too)
```suggestion
```

File: esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include "esphome/core/component.h"
+#include "esphome/components/sensor/sensor.h"
+#include "esphome/components/esp32_ble_tracker/esp32_ble_tracker.h"
+#include "esphome/components/xiaomi_ble/xiaomi_ble.h"
```
Comment (jesserockz) [inline esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.h]: Not used
```suggestion
```

File: esphome/components/xiaomi_hhccjcy10/sensor.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await esp32_ble_tracker.register_ble_device(var, config)
+
+    cg.add(var.set_address(config[CONF_MAC_ADDRESS].as_hex))
+
+    if CONF_TEMPERATURE in config:
+        sens = await sensor.new_sensor(config[CONF_TEMPERATURE])
+        cg.add(var.set_temperature(sens))
```
Comment (jesserockz) [inline esphome/components/xiaomi_hhccjcy10/sensor.py]: Please alter these to use this method of checking
```suggestion
    if temperature_config := config.get(CONF_TEMPERATURE):
        sens = await sensor.new_sensor(temperature_config)
        cg.add(var.set_temperature(sens))
```

File: esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      ESP_LOGVV(TAG, "no tuya service data UUID.");
+      continue;
+    }
+    if (service_data.data.size() != 9) {  // tuya alternate between two service data
+      continue;
+    }
+    const uint8_t *data = service_data.data.data();
+
+    const int16_t temperature = encode_uint16(data[1], data[2]);
+    this->temperature_->publish_state((float) temperature / 10.0f);
```
Comment (jesserockz) [inline esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.cpp]: These need to be checked that the user actually configured them
```suggestion
    if (this->temperature_ != nullptr) {
      const int16_t temperature = encode_uint16(data[1], data[2]);
      this->temperature_->publish_state((float) temperature / 10.0f);
    }
```

File: esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+#ifdef USE_ESP32
+
+namespace esphome {
+namespace xiaomi_hhccjcy10 {
+
+class XiaomiHHCCJCY10 : public Component, public esp32_ble_tracker::ESPBTDeviceListener {
+ public:
+  void set_address(uint64_t address) { address_ = address; }
```
Comment (jesserockz) [inline esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.h]: Member variables and function calls need to all be prefixed with `this->`
```suggestion
  void set_address(uint64_t address) { this->address_ = address; }
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.h (hunk 5 above)
Comment (fariouche) [reply esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.h]: done

File: esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.cpp (hunk 4 above)
Comment (fariouche) [reply esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.cpp]: done

File: esphome/components/xiaomi_hhccjcy10/sensor.py (hunk 3 above)
Comment (fariouche) [reply esphome/components/xiaomi_hhccjcy10/sensor.py]: done

File: esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.h (hunk 2 above)
Comment (fariouche) [reply esphome/components/xiaomi_hhccjcy10/xiaomi_hhccjcy10.h]: done

File: esphome/components/xiaomi_hhccjcy10/sensor.py (hunk 1 above)
Comment (fariouche) [reply esphome/components/xiaomi_hhccjcy10/sensor.py]: done

**Comment by fariouche** [comment]
Hello, any news?

**Review by jesserockz — APPROVED** [review]

**Pull request merged**